File: capture/android/PressurePulsationsRecorder/src/main/java/com/avyss/PressurePulsationsRecorder/acquisition/AbstractSampleCollector.java
package com.avyss.PressurePulsationsRecorder.acquisition;

import android.util.Log;

abstract public class AbstractSampleCollector {

    private static final long NANOS_IN_SECOND = 1000*1000*1000;

    final private float samplesPerSecond;
    final private int   maxSamples;
    final private long  recStartTimeNanos;

    private volatile boolean collectionFinished;
    private long firstSampleTimeNanos;
    private int  lastSampleIndex;
    private int   [] counts;
    private float [] sums;

    public AbstractSampleCollector(float samplesPerSecond, int maxRecordingLengthSec, long recStartTimeNanos) {

        this.samplesPerSecond = samplesPerSecond;
        this.maxSamples = Math.round(maxRecordingLengthSec * samplesPerSecond);
        this.recStartTimeNanos = recStartTimeNanos;

        firstSampleTimeNanos = -1;
        lastSampleIndex = -1;

        counts = new int[maxSamples];
        sums = new float[maxSamples];

        collectionFinished = false;
    }

    public void stopCollecting() {
        collectionFinished = true;
    }

    public float [] getCollectedSamples() {

        if (!collectionFinished) {
            throw new IllegalStateException("samples are still being collected, call stopCollecting() first");
        }

        // no samples captured?
        if (lastSampleIndex < 0) {
            return new float[0];
        }

        // some samples were captured
        float [] samples = new float[lastSampleIndex + 1];

        for (int i = 0; i < lastSampleIndex + 1; i++) {
            if (counts[i] == 0) {
                samples[i] = Float.NaN;
            } else {
                float currValue = sums[i] / counts[i];
                samples[i] = currValue;
            }
        }

        return samples;
    }

    public float firstSampleTimeDelay() {
        if (!collectionFinished) {
            throw new IllegalStateException("samples are still being collected, call stopCollecting() first");
        }

        // no samples captured?
        if (lastSampleIndex < 0) {
            return Float.NaN;
        }

        // some samples were captured
        return (float) (firstSampleTimeNanos - recStartTimeNanos) / NANOS_IN_SECOND;
    }

    public float getSamplingRate() {
        return samplesPerSecond;
    }

    protected Float onSampleAcquired(float sampleValue, long timestamp) {

        if (collectionFinished) {
            return null;
        }

        if (firstSampleTimeNanos == -1) {
            firstSampleTimeNanos = timestamp;
        }

        float elapsedSeconds = (float)(timestamp - firstSampleTimeNanos) / NANOS_IN_SECOND;

        int sampleIndex = Math.round(elapsedSeconds * samplesPerSecond);

        Log.v("received", "index " + sampleIndex + " value " + sampleValue);

        if (sampleIndex < 0 || sampleIndex >= maxSamples) {
            return null;
        }

        sums[sampleIndex] += sampleValue;
        counts[sampleIndex]++;

        lastSampleIndex = sampleIndex;
        return elapsedSeconds;
    }

}

File: capture/android/PressurePulsationsRecorder/src/main/java/com/avyss/PressurePulsationsRecorder/acquisition/AbstractSampleCollector.kt
package com.avyss.PressurePulsationsRecorder.acquisition

import android.util.Log

abstract class AbstractSampleCollector(
        private val nValuesPerSample: Int,
        val samplingRate: Float,
        maxRecordingLengthSec: Int,
        private val recStartTimeNanos: Long
) {

    companion object {
        private const val NANOS_IN_SECOND: Long = 1000L * 1000L * 1000L
    }

    @Volatile
    private var collectionFinished: Boolean = false

    private val maxNSamples = Math.round(maxRecordingLengthSec * samplingRate)
    private val counts      = IntArray(maxNSamples)
    private val sums        = Array(nValuesPerSample) {FloatArray(maxNSamples)}

    val collectedData: Array<FloatArray>
        get() {

            if (!collectionFinished) {
                throw IllegalStateException("samples are still being collected, call stopCollecting() first")
            }

            // find index of first actual sample
            var idx = 0;
            while (idx < maxNSamples && counts[idx] == 0) {
                idx++;
            }
            if (idx == maxNSamples) {
                return Array(0, {FloatArray(0)})
            }
            val firstSampleIdx = idx;

            // find index of last actual sample
            // (it may be the same one as the first, if there only one sample has arrived)
            var lastSampleIdx = firstSampleIdx;
            while (idx < maxNSamples) {
                if (counts[idx] != 0) {
                    lastSampleIdx = idx;
                }
                idx++;
            }

            val retData = Array(1 + nValuesPerSample, {FloatArray(lastSampleIdx - firstSampleIdx + 1)})

            for (i in firstSampleIdx .. lastSampleIdx) {
                val time = i.toFloat() / samplingRate
                retData[0][i] = time
                for (n in 0 until nValuesPerSample) {
                    if (counts[i] == 0) {
                        retData[n+1][i] = java.lang.Float.NaN
                    } else {
                        retData[n+1][i] = sums[n][i] / counts[i]
                    }
                }
            }

            return retData
        }

    fun stopCollecting() {
        collectionFinished = true
    }

    protected fun onSampleAcquired(timestamp: Long, vararg sampledValues: Float) {

        if (collectionFinished) {
            return
        }

        if (sampledValues.size != nValuesPerSample) {
            throw IllegalArgumentException()
        }

        val elapsedSeconds = (timestamp - recStartTimeNanos).toFloat() / NANOS_IN_SECOND

        val sampleIndex = Math.round(elapsedSeconds * samplingRate)

        Log.v("received", "index $sampleIndex value " + sampledValues.joinToString { f->f.toString() })

        if (sampleIndex < 0 || sampleIndex >= maxNSamples) {
            return
        }

        for (n in 0 until nValuesPerSample) {
            sums[n][sampleIndex] += sampledValues[n]
        }

        counts[sampleIndex]++
    }

}
